script.module.cache/lib/cache.py

```diff
@@ -0,0 +1,311 @@
+# -*- coding: utf-8 -*-
+
+"""HTTP File cache and helpers"""
+
+__author__ = "fraser"
+
+import json
+import logging
+import sqlite3
+import time
+from datetime import datetime, timedelta, tzinfo
+from os import path
+
+import xbmc
+import xbmcaddon
+import xbmcvfs
+
+try:
+    import _pickle as cpickle
+except ImportError:
+    import cPickle as cpickle
+
+logger = logging.getLogger(__name__)
+
+HTTP_DATE_FORMAT = "%a, %d %b %Y %H:%M:%S %Z"
+ADDON = xbmcaddon.Addon()
+ADDON_PROFILE = ADDON.getAddonInfo("profile")
+CACHE_URI = xbmc.translatePath(path.join(ADDON_PROFILE, "cache.sqlite"))
+
+
+def httpdate_to_datetime(input_date):
+    # type: (str) -> Union[datetime, None]
+    """Converts http date string to datetime"""
+    if input_date is None:
+        return None
+    try:
+        return datetime.strptime(input_date, HTTP_DATE_FORMAT)
+    except TypeError:
+        return datetime(*(time.strptime(input_date, HTTP_DATE_FORMAT)[0:6]))
+    except ValueError as e:
+        logger.debug(e)
+        return None
+
+
+def datetime_to_httpdate(input_date):
+    # type: (datetime) -> Union[str, None]
+    """Converts datetime to http date string"""
+    if input_date is None:
+        return None
+    try:
+        return input_date.strftime(HTTP_DATE_FORMAT)
+    except (ValueError, TypeError) as e:
+        logger.debug(e)
+        return None
+
+
+def conditional_headers(row):
+    # type: (sqlite3.Row) -> dict
+    """Creates dict of If-None-Match and If-Modified-Since headers based on row etag and last_modified columns"""
+    headers = {}
+    if row["etag"] is not None:
+        headers["If-None-Match"] = row["etag"]
+    if row["last_modified"] is not None:
+        headers["If-Modified-Since"] = datetime_to_httpdate(row["last_modified"])
+    return headers
+
+
+class Store(object):
+    """
+    Generic unique string storage helper
+    Saves unique strings in a list that can
+    be retrieved, appended to, removed from or cleared entirely
+    """
+
+    def __init__(self, key, db=None):
+        # type: (str, str) -> None
+        """
+        Creates a new Store object
+        key is the identifier that strings are stored under
+        db is the path to a sqlite3 database (defaults to Cache default)
+        """
+        self.db = db
+        self.key = key
+
+    def retrieve(self):
+        # type: () -> list
+        """Gets stored strings as list"""
+        with Cache(self.db) as c:
+            data = c.get(self.key)
+            return json.loads(data["blob"]) if data else []
+
+    def _save(self, data):
+        # type: (list) -> None
+        """Saves stored data"""
+        with Cache(self.db) as c:
+            c.set(self.key, json.dumps(data, ensure_ascii=False))
+
+    def append(self, data):
+        # type: (str) -> None
+        """Add string to the store if it doesn't exist"""
+        current = self.retrieve()
+        if data not in current:
```

Comment:
If strings in the list are unique, consider using a set instead of a list. Presence check in a set is an O(1) operation vs O(n) in a list. The only problem with set is that it does not maintain the element's order.

Reply:
Ah my thinking was sets aren't supported by JSON...so a list kind of made sense as it can be passed to json.loads/json.dumps ... but thinking about it now - the data doesn't need serialising further at all as a pickled blob of a set is fine...indeed this removes the dependency on json too - done and done :)